File: plugin/src/main/groovy/com/shogo82148/ribbonizer/filter/GrayScaleFilter.java
package com.shogo82148.ribbonizer.filter;

import com.shogo82148.ribbonizer.resource.Filter;
import com.shogo82148.ribbonizer.resource.ImageAdaptiveIcon;
import com.shogo82148.ribbonizer.resource.ImageIcon;
import com.shogo82148.ribbonizer.resource.Resource;

import java.awt.image.BufferedImage;
import java.util.function.Consumer;

public class GrayScaleFilter implements Consumer<Resource>, Filter {

    private static int toGray(int color) {
        int a = (color & 0xFF000000);
        int r = (color & 0x00FF0000) >> 16;
        int g = (color & 0x0000FF00) >> 8;
        int b = (color & 0x000000FF);

        int c = (int) ((2.0 * r + 4.0 * g + b) / 7.0);
        return a | (c << 16) | (c << 8) | c;
    }

    @Override
    public void accept(Resource rsc) {
        rsc.apply(this);
    }

    public void apply(ImageIcon icon) {
        BufferedImage image = icon.getImage();
        int width = image.getWidth();
        int height = image.getHeight();

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int color = image.getRGB(x, y);
                image.setRGB(x, y, toGray(color));
            }
        }
    }

    public void apply(ImageAdaptiveIcon icon) {
        BufferedImage image = icon.getImage();
        int width = image.getWidth();
        int height = image.getHeight();

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int color = image.getRGB(x, y);
                image.setRGB(x, y, toGray(color));
            }
        }
    }

}

File: plugin/src/main/java/com/shogo82148/ribbonizer/filter/GrayScaleFilter.kt
package com.shogo82148.ribbonizer.filter

import com.shogo82148.ribbonizer.resource.Filter
import com.shogo82148.ribbonizer.resource.ImageAdaptiveIcon
import com.shogo82148.ribbonizer.resource.ImageIcon
import com.shogo82148.ribbonizer.resource.Resource
import java.util.function.Consumer

class GrayScaleFilter : Consumer<Resource>, Filter {
    override fun accept(rsc: Resource) {
        rsc.apply(this)
    }

    override fun apply(icon: ImageIcon) {
        val image = icon.image
        val width = image.width
        val height = image.height
        for (y in 0 until height) {
            for (x in 0 until width) {
                val color = image.getRGB(x, y)
                image.setRGB(x, y, toGray(color))
            }
        }
    }

    override fun apply(icon: ImageAdaptiveIcon) {
        val image = icon.image
        val width = image.width
        val height = image.height
        for (y in 0 until height) {
            for (x in 0 until width) {
                val color = image.getRGB(x, y)
                image.setRGB(x, y, toGray(color))
            }
        }
    }

}

private fun toGray(color: Int): Int {
    val a = (color and 0xFF000000.toInt())
    val r = (color and 0x00FF0000) shr 16
    val g = (color and 0x0000FF00) shr 8
    val b = (color and 0x000000FF)
    val c = ((2.0 * r + 4.0 * g + b) / 7.0).toInt()
    return a or (c shl 16) or (c shl 8) or c
}
